atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java

```diff
@@ -164,20 +183,189 @@ private long storeAndCheckExpiringByteStreams(int size) throws IOException {
+                deletingStreamStore.deleteStreams(t, ImmutableSet.of(streamId));
+            }
+        });
+
+        assertStreamDoesNotExist(streamId);
+    }
+
+    @Test
+    public void testStreamMetadataConflictWriteFirst() throws Exception {
+        long streamId = timestampService.getFreshTimestamp();
+
+        runConflictingTasksConcurrently(streamId, new TwoConflictingTasks() {
+            @Override
+            public void startFirstAndFail(Transaction t, long streamId) {
+                DeletingStreamStore deletingStreamStore = new DeletingStreamStore(StreamTestStreamStore.of(txManager, StreamTestTableFactory.of()));
+                deletingStreamStore.deleteStreams(t, ImmutableSet.of(streamId));
+            }
+
+            @Override
+            public void startSecondAndFinish(Transaction t, long streamId) {
+                StreamTestStreamStore ss = StreamTestStreamStore.of(txManager, StreamTestTableFactory.of());
+                ss.storeStreams(t, ImmutableMap.of(streamId, new ByteArrayInputStream(new byte[1])));
+            }
+        });
+
+        assertNotNull(getStream(streamId));
+    }
+
+    private InputStream getStream(long streamId) {
+        return txManager.runTaskThrowOnConflict(t -> {
+            StreamTestStreamStore streamStore = StreamTestStreamStore.of(txManager, StreamTestTableFactory.of());
+            return streamStore.loadStream(t, streamId);
+        });
+    }
+
+    private void assertStreamDoesNotExist(final long streamId) {
+        try {
+            getStream(streamId);
+            fail("This element should have been deleted");
+        } catch (NoSuchElementException e) {
+            // expected
+        }
+    }
+
+    private void runConflictingTasksConcurrently(long streamId, TwoConflictingTasks twoConflictingTasks) throws InterruptedException {
+        final CountDownLatch firstLatch = new CountDownLatch(1);
+        final CountDownLatch secondLatch = new CountDownLatch(1);
+
+        ExecutorService exec = PTExecutors.newFixedThreadPool(2);
+
+
+        Future<?> firstFuture = exec.submit(() -> {
+            try {
+                txManager.runTaskThrowOnConflict(t -> {
+                    twoConflictingTasks.startFirstAndFail(t, streamId);
+                    letOtherTaskFinish(firstLatch, secondLatch);
+                    return null;
+                });
+                fail("Because we concurrently wrote, we should have failed with TransactionConflictException.");
+            } catch (TransactionConflictException e) {
+                // expected
+            }
+        });
+
+        firstLatch.await();
+
+        Future<?> secondFuture = exec.submit((Runnable) () -> txManager.runTaskThrowOnConflict((TransactionTask<Void, RuntimeException>) t -> {
+            twoConflictingTasks.startSecondAndFinish(t, streamId);
+            return null;
+        }));
+
+        exec.shutdown();
```

Comment:
I think you want to also shutdown in finally block in case this test fails to avoid flaking others